servers/movie_writer/movie_writer.h

```diff
@@ -0,0 +1,75 @@
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef MOVIE_WRITER_H
+#define MOVIE_WRITER_H
+
+#include "core/templates/local_vector.h"
+#include "servers/audio/audio_driver_dummy.h"
+#include "servers/audio_server.h"
+
+class MovieWriter : public Object {
+	GDCLASS(MovieWriter, Object);
+
+	uint64_t fps = 0;
+	uint64_t mix_rate = 0;
+	uint32_t audio_channels = 0;
+
+	LocalVector<int32_t> audio_mix_buffer;
+
+	static MovieWriter *writer;
+
+protected:
+	virtual uint32_t get_audio_mix_rate() const;
+	virtual AudioServer::SpeakerMode get_audio_speaker_mode() const;
+
+	virtual Error write_begin(const Size2i &p_movie_size, uint32_t p_fps, const String &p_base_path);
+	virtual Error write_frame(const Ref<Image> &p_image, const int32_t *p_audio_data);
+	virtual void write_end();
+
+	GDVIRTUAL0RC(uint32_t, _get_audio_mix_rate)
+	GDVIRTUAL0RC(AudioServer::SpeakerMode, _get_audio_speaker_mode)
+
+	GDVIRTUAL3R(Error, _write_begin, const Size2i &, uint32_t, const String &)
+	GDVIRTUAL2R(Error, _write_frame, const Ref<Image> &, GDNativeConstPtr<int32_t>)
+	GDVIRTUAL0(_write_end)
+
+	static void _bind_methods();
+
+public:
+	static void register_writer(MovieWriter *p_writer);
+	static MovieWriter *get_registered_writer();
+
+	void begin(const Size2i &p_movie_size, uint32_t p_fps, const String &p_base_path);
```

Comment:
I would add a way to select codec / get list of supported codes directly to the API (for the future implementations):
```suggestion
	void begin(const Size2i &p_movie_size, uint32_t p_fps, const String &p_base_path, const String &p_codec = String());
	PackedStringArray get_supported_codec_list() const;
```
I guess, at some point, it's probably worth to implement support for the platform specific OS built-in H264/HEVC hardware encoders (as module or GDExtension).

Also, what's intended way to select active `MoveWriter`, I see only `register_writer` but it seems to have only one writer saved, and there's no command line argument or UI to select which one to use.

Reply:
I did not want to make it super flexible, but if there is demand for it we could make it possible to use more than one.

Reply:
A more flexible API that supports multiple codecs is desired. MJPEG is old and was beginning to be replaced by h264 in 2004. It will be used if it's the only option, but will be replaced as soon as a better option is available. If it could pipe to ffmpeg, I would immediately go with png if I needed transparency, or x264 if I wanted higher quality and lower size, especially if it was a final deliverable.